lite: enable (u)int8 quantization and int32 for ABS

zhenhuaw-me opened the pull request:
Hi, this PR adds uint8/int8/int32 support for ABS kernel in TFLite. Detail change includes:
1. Move ABS out routine and kernel from elementwise generic as ops have different quantization handling.
2. Float/Int32 ABS invoke `std::abs` directly.
3. Int8/UInt8 ABS involving requantization after `std::abs`.
4. Float as version 1, Int32/Int8/UInt8 as version 2.
5. Test for kernel and TOCO.

jackwish commented:
One open discussion is that, do we really need requantization in quantized ABS.

Intuitively, it's not needed since the representation of FP32 in INT8 won't change, that is the OP's property. However, we can safely remove the requantize only when TOCO enforcing the same quantization representation for input and output tensor, which seems is not the case currently. We have a model which has different quantization representation of ABS input/output when converted to TFLite model. So maybe, we can investigate that should TOCO (and any similar quantization tools) enforce such check when converting models.

Would you please help to CC people who may have interest or expertise in such topic (I mean where need to evaluate quantization representation handling in model conversion)? Looking forward to hear from you guys. Thank you.

zhenhuaw-me committed:
lite: enable (u)int8 quantization and int32 for ABS

1. Move ABS out routine and kernel from elementwise
   generic as ops have different quantization handling.
2. Float/Int32 ABS invoke `std::abs` directly.
3. Int8/UInt8 ABS involving requantization after `std::abs`.
4. Float as version 1, Int32/Int8/UInt8 as version 2.
5. Test for kernel and TOCO.

Test:
bazel run //tensorflow/lite/kernels:abs_test
bazel run //tensorflow/lite/toco/tflite:operator_test

zhenhuaw-me committed:
format code with clang-format of f4eb2fa6dd

Some are formatting are not read friendly, have another commit to
discuss whether need such changes.

jackwish commented:
Hi there, any update needed?

jackwish commented:
Two more commits to extend Quantization-aware Training to operator Sub, Abs, Squeeze, ExpandDims and so on; and remove requantize in Abs TFLite kernel as it should have same quantization representation.

Also loop @aselle and @suharshs who may have interest.

suharshs commented on tensorflow/lite/kernels/abs.cc:
```diff
@@ -0,0 +1,136 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
nit: could you make these 2019

jackwish replied:
Sure, thank you!

suharshs commented on tensorflow/lite/kernels/internal/types.h:
```diff
@@ -758,6 +758,12 @@ struct ArithmeticParams {
 };
 
+struct AbsParams {
+  // uint8 inference params.
```
uint8/int8

jackwish replied:
Yes, will update.

suharshs commented on tensorflow/lite/kernels/test_util.h:
```diff
@@ -65,6 +65,12 @@ inline std::vector<float> Dequantize(const std::vector<T>& data, float scale,
 }
 
+float GetQuantizeTolerance(int min, int max) {
+  float QuantizedStep = (max - min) / 255.0;
```
nit: Name QuantizedStep quantized_step and same for other names.

jackwish replied:
Yes, will update!

suharshs commented on tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc:
```diff
@@ -430,6 +430,7 @@ ::tensorflow::Status HardcodeMinMax::Run(Model* model, std::size_t op_index,
       changed = HardcodeMinMaxForAverageOrMaxPool(model, op);
       break;
 
+    case OperatorType::kAbs:
```
is this hardcode case intended? This hardcodes the min/max of the output to the min/max of the input. this means if the abs input is [-2, -1] than the output range will be the same, which seems off. 

Couple options:
(1) Choose to not have a hardcode case here, instead relying on a Fakequant from contrib/quatnize?
(2) Hardcode to:
r = max(abs(input_min), (input_max))
[-r, r]

WYDT?

jackwish replied:
Yes, it **was** intended. Like mentioned https://github.com/tensorflow/tensorflow/pull/31914#issuecomment-524169853, I have been choosing whether to let input/output of Abs having different quantization representation. But I really missed the value range that you pointed out. This is not what we want.

jackwish replied:
Removed Abs from hardcode, maybe we can learn min/max in further patch.

suharshs commented on tensorflow/lite/toco/tflite/operator_test.cc:
```diff
@@ -158,6 +158,37 @@ TEST_F(OperatorTest, SimpleOperators) {
+  Model float_model;
+  Array& input_float_array = float_model.GetOrCreateArray(abs_op.inputs[0]);
+  input_float_array.data_type = ArrayDataType::kFloat;
+  OperatorSignature float_signature = {.op = &abs_op, .model = &float_model};
```
unfortunately this .op notation breaks on come compilers :(

jackwish replied:
Ok then.. I was copying from some other tests...

suharshs commented on tensorflow/lite/toco/tflite/operator.cc:
```diff
@@ -182,6 +182,22 @@ class DepthwiseConvolution
+    // Version 2 supports signed/unsigned int8 and signed int32 input types.
+    if (input_array.data_type == ArrayDataType::kInt8 ||
+        input_array.data_type == ArrayDataType::kUint8 ||
+        input_array.data_type == ArrayDataType::kInt32) {
```
thanks! please update version in op resolver too https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/register.cc#L161

jackwish replied:
Sure. I am thinking that do we need a guide to let people know the *must* of adding operators related code to TFLite? I believe many people have similar questions like  https://github.com/tensorflow/tensorflow/issues/31879

suharshs commented on tensorflow/lite/toco/tflite/operator.cc:
```diff
@@ -1060,6 +1076,8 @@ class Lstm : public BuiltinOperator<LstmCellOperator, ::tflite::LSTMOptions,
         // KERNEL_BASIC was added in version 2.
         return 2;
+      default:
+        return -1;
```
this was intentionally left out to have a compiler error if we miss an case

jackwish replied:
Got it, I will leave some comments here :)

suharshs commented on tensorflow/contrib/quantize/python/quantize.py:
```diff
@@ -37,10 +37,10 @@
 _QUANTIZATION_OP = {'FakeQuantWithMinMaxVars'}
-_VALID_SRC_OP = {'Add', 'AddV2', 'Mul'}
-_INTERMEDIATE_OP = {'Add', 'AddV2', 'Mul'}
+_VALID_SRC_OP = {'Add', 'AddV2', 'Mul', 'Sub', 'ConcatV2'}
```
could we add a test for the contrib/quantize changes, and move the contrib/quantize changes to a separate pr from the tflite changes.

jackwish replied:
Sure, I will add some tests for this.

Regarding moving the `contrib/quantize` out, I am not sure about setting min/max of ABS in TOCO, which should be consistent with the TFLite kernel I think.

jackwish replied:
Removed `contrib/quantize` changes.

suharshs reviewed (commented):
Awesome work!

suharshs commented:
I will spend some time to think more about the requantization case for abs, for now i think going to int32 like you are doing is fine. Thanks!

jackwish commented:
> I will spend some time to think more about the requantization case for abs, for now i think going to int32 like you are doing is fine. Thanks!

Thank you for the review @suharshs !  For the Abs requantization, I was paying too much attention on the representation precision about input/output - for most case overlapping zero, the output range is smaller than the input range, but the data representation cannot be more accurate in output. I mean, the information of output cannot be more than input, so I thought a requantization is not needed. Leading to miss the range case `[-2, -1]` you pointed out, for which a requantization seems needed. Thank you for the patient review!

The min/max design choice also impacts `contrib/quantize`, where I took ABS as *pass through* operators - while should not. I think learn min/max for Abs rather than hardcode it is a good idea, just like what was done for Concat in this PR.

Also, out of the scope for this PR, I am thinking that the open sourced Quantization-aware Training may need redesign. Current design basically takes operators which holding weights as something like *master* node, and processes based on them. Many more operators, such as the already added `Add` and `Mul`, can be improved by QAT. I guess we can add FakeQuant to anywhere, unless it is not necessary - `Reshape` for example. We may let TensorFlow operators holding some attributions describing where the quantization representation of their inputs/outputs may be different, and the `contrib/quantize` can simply walk the network and insert FakeQuant nodes. That is what I am thinking, as someone who don't hold a full sight of the design. I will be very happy to contribute such things. Hoping it were not going too far wrong. :)

I will update the PR next week to let Abs learn min/max. I am thinking that the `contrib/quantize` and tflite part code can be in one PR to let them have consistent behavior on quantization related things (int32 is really insignificant). What do you think?

jackwish commented:
Removed `contrib/quantize` changes from this PR. Will open another for it. Please review again @suharshs .

zhenhuaw-me committed:
address Suharsh's review comments

zhenhuaw-me committed:
lite: let the map key of op version in A-Z style

Such that we can find easily and avoid duplication.

Test:
bazel run //tensorflow/lite/kernels/abs_test
bazel run //tensorflow/lite/toco/tflite/operator_test
bazel run //tensorflow/lite/tools/versioning/op_version_test

gbaned commented:
@jackwish Could you please check build failures and resolve the conflicts? Thanks!

jackwish commented:
> @jackwish Could you please check build failures and resolve the conflicts? Thanks!

Hi @gbaned , conflicts has been fixed, thank you for the reminder.
Hi @suharshs , could you please have another look in this?

zhenhuaw-me committed:
lite: fix conflicts regarding new op versioning

Which was introduced in c8d3be804a.
Also remove duplication of ABS in op version.

Test:
bazel run //tensorflow/lite/kernels/abs_test
bazel run //tensorflow/lite/toco/tflite/operator_test
bazel run //tensorflow/lite/tools/versioning/op_version_test

gbaned commented:
@jackwish Could you please resolve the conflicts? Thanks!

jackwish commented:
Hi @gbaned the conflicts have been fixed. The latest two rounds of conflict resolving are due to the long progress of this PR. Hoping that we can accelerate the review progress before new conflicts emerge :)

gbaned commented:
@jackwish Sorry for the slow response. Could you please resolve the conflicts? Thanks!

gbaned commented:
@jackwish Could you please check build failures and resolve the conflicts? Thanks!

jackwish commented:
Hi @gbaned the conflicts have been fixed, and @suharshs @wangtz @jianlijianli please have another look.

PS. As tree round of conflicts resolving has been conducted in this PR, I am now unsubscribing from this PR thread and leaving it as it is. Thank you guys for all the effort, best regards.

gbaned commented:
@jackwish Could you please resolve the conflicts? Thanks!

jackwish commented:
> @jackwish Could you please resolve the conflicts? Thanks!

Thank you @gbaned for the remind.

But the conflicts resolving will not happen. As https://github.com/tensorflow/tensorflow/pull/31914#issuecomment-544493621 has stated, this PR is in a *conflicts resolving* loop which I will no longer pay efforts. Please feel free to close this PR if no further actions are going to happen from your side.

Thank you everyone for your time.

jackwish commented:
Now closing this PR. Thanks everyone for your time.

I will save the time of contributing back to tensorflow official repo to maintain a quantization enhanced fork based on TensorFlow 1.14. For anyone interested in this PR or any further related features, please switch to [my repo](https://github.com/jackwish/tensorflow/tree/lts/r1.14/).

The enhancement covers two aspects basically:
* More quantized operators and kernels in TensorFlow Lite.
* More scenarios for Quantization-aware Training.

jackwish commented:
switch to #34263
